### PR TITLE
[Bug fix] Fix optional_output_tensor handling in ttnn.mse_loss and ttnn.l1_loss for SUM/MEAN reductions

### LINKED ISSUE #56295: [ops/ttnn]: ttnn.mse_loss & ttnn.l1_loss crash or silently corrupt preallocated output_tensor on SUM/MEAN reduction
### Component / Area

ops, ttnn

### Issue Type (optional)

Bad Outputs

### Observed

In `ttnn/cpp/ttnn/operations/loss/loss.cpp`, `loss_function` accepts an `optional_output_tensor` parameter. When `reduce_mode` is `LossReductionMode::SUM` or `LossReductionMode::MEAN`, `optional_output_tensor` is forwarded directly to `ttnn::subtract` before reduction is applied:

1. Runtime Crash (Shape Mismatch): If the caller preallocates `output_tensor` matching the expected scalar/tile output of reduced loss (e.g. `[1, 1, 32, 32]`), `ttnn::subtract` crashes with a fatal shape mismatch against input `[B, C, H, W]`.
2. Silent Data Corruption: If the caller preallocates `output_tensor` with the input shape `[B, C, H, W]`, `ttnn::subtract` populates it with intermediate unreduced elementwise squared/absolute differences, while `ttnn::sum`/`ttnn::mean` allocates a brand new tensor for the return value. The caller's preallocated `output_tensor` is never populated with the reduced loss and is silently left holding intermediate difference values.

### Expected

`output_tensor` should be correctly populated with the final reduced scalar/loss value (via `ttnn::copy`) when reduction is `SUM` or `MEAN`, matching the returned tensor.

### 1. Steps (exact commands)

```python
import torch
import ttnn

device = ttnn.open_device(0)

shape = [1, 1, 64, 64]
ref = ttnn.from_torch(torch.randn(shape, dtype=torch.bfloat16), layout=ttnn.TILE_LAYOUT, device=device)
pred = ttnn.from_torch(torch.randn(shape, dtype=torch.bfloat16), layout=ttnn.TILE_LAYOUT, device=device)

# Case 1: Preallocated scalar output crashes with shape mismatch
out_scalar = ttnn.empty([1, 1, 32, 32], dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
# Crashes in ttnn::subtract:
res = ttnn.mse_loss(ref, pred, reduction=ttnn.LossReductionMode.MEAN, output_tensor=out_scalar)

# Case 2: Preallocated full shape buffer suffers silent corruption
out_buffer = ttnn.empty(shape, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
res2 = ttnn.mse_loss(ref, pred, reduction=ttnn.LossReductionMode.MEAN, output_tensor=out_buffer)
# out_buffer contains unreduced elementwise difference, NOT the mean loss!
```

### 2. Input data / link or description

Synthetic random tensors in BFLOAT16 with TILE_LAYOUT on device.

### 3. Frequency

Always (Deterministic 100%)

### 1. Software Versions

OS: Ubuntu 22.04 / Any
Branch: main (HEAD)

### 2. Hardware Details

All architectures (Wormhole / Blackhole)

### Is this a regression?

Unknown

### Regression Details

_No response_

### Root Cause Analysis

In `ttnn/cpp/ttnn/operations/loss/loss.cpp`:
```cpp
Tensor result = ttnn::subtract(ref, prediction, std::nullopt, memory_config, optional_output_tensor, fused_ops);
```

`optional_output_tensor` should only be passed to `ttnn::subtract` when `reduce_mode == LossReductionMode::NONE`. For `SUM` and `MEAN`, subtraction should write to an intermediate temporary tensor, and after reduction, `ttnn::copy` should copy the reduced result into `optional_output_tensor.value()`.

---

### Ready-to-Submit Fix & Bounty Consideration

I have verified this root cause, formatted a clean C++ fix using clang-format, and added regression test coverage in `tests/ttnn/nightly/unit_tests/operations/eltwise/test_losses.py`.

If this looks good to the maintainers, I would be glad to submit the Pull Request immediately!

Additionally, given that this addresses silent data corruption and runtime crashes in core loss operations, could this finding be considered eligible for the Tenstorrent Community Bounty program (@jberkowitzTT)?

Either way, happy to contribute the fix and help improve tt-metal!

### Priority

P1

### Impact

_No response_

### PR DESCRIPTION
**Pull request by MusounoEnma**
### PR Category
Bug fix

### Summary
Fixes #56295

In `ttnn/cpp/ttnn/operations/loss/loss.cpp`, `loss_function` previously passed `optional_output_tensor` unconditionally to `ttnn::subtract`. When `reduce_mode` is `LossReductionMode::SUM` or `LossReductionMode::MEAN`:
1. If the caller preallocated `output_tensor` with the reduced scalar/tile shape `[1, 1, 32, 32]`, `subtract` crashed with a fatal shape mismatch against input `[B, C, H, W]`.
2. If the caller preallocated `output_tensor` with full input shape `[B, C, H, W]`, `subtract` wrote unreduced elementwise differences into it, while `mean`/`sum` allocated a brand new tensor for the return value, silently corrupting the caller's preallocated buffer.

**Changes:**
- In `loss_function`, only pass `optional_output_tensor` to `ttnn::subtract` when `reduce_mode == LossReductionMode::NONE`.
- For `SUM` and `MEAN`, compute subtraction into an intermediate tensor, perform the reduction, and copy the reduced result into `optional_output_tensor` via `ttnn::copy`.
- Added official regression tests in `tests/ttnn/nightly/unit_tests/operations/eltwise/test_losses.py` (`test_mse_loss_with_output_tensor`, `test_l1_loss_with_output_tensor`) covering `NONE`, `MEAN`, and `SUM` reductions.

### Notes for reviewers
- Code formatted with `clang-format`.
- Fix adheres to standard `ttnn::copy` semantics for preallocated device buffers.
- Added regression tests for both `mse_loss` and `l1_loss` with preallocated `output_tensor`.

### CODE REVIEW

**Comment by MusounoEnma** [comment]
Hi @bbradelTT @mateusznowakTT,

Just checking in on this PR when you have a chance to take a look.

The changes handle `optional_output_tensor` properly via `ttnn::copy` for reduced loss modes, and regression tests for both `mse_loss` and `l1_loss` across `NONE`, `MEAN`, and `SUM` reductions are included and passing.

Please let me know if any tweaks or additional tests are needed. Thank you!

**Comment by MusounoEnma** [comment]
Hi @mateusznowakTT @bbradelTT, bumping this as requested reviewers are currently unassigned. All CI tests and regression suites for optional_output_tensor on ttnn.mse_loss/ttnn.l1_loss (SUM/MEAN) are passing. Could you please take a look when you have a moment?
